### PR TITLE
Ensure monitor captures the deletion

### PR DESCRIPTION
**Pull request by yehornaumenko**
### Which issue this PR addresses:

https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/12371799/

### What this PR does / why we need it:

[Google doc](https://docs.google.com/document/d/1x6VPMWaD_hADHB3UOsksSwc7dJAuzbd5LhekFXgRXkc/edit?usp=sharing)

### Is there any documentation that needs to be updated for this PR?
`docs/monitoring.md`

### CODE REVIEW

**Comment by jhoreman** [comment]
Hey Yehor, thanks for this. Can we include the work item associated with this PR? Would like to know more about the reasoning and goals behind this change.

**Comment by yehornaumenko** [comment]
Hi @jhoreman. I have updated the description of this PR, could you check it, please?

**Comment by jhoreman** [comment]
> Hi @jhoreman. I have updated the description of this PR, could you check it, please?

Perfect! Thanks for adding all this detail.

File: pkg/backend/openshiftcluster.go [hunk 1]
```diff
@@ -185,6 +186,11 @@ func (ocb *openShiftClusterBackend) handle(ctx context.Context, log *logrus.Entr
 
 		stop()
 
+		// This Sleep ensures that the monitor has enough time
+		// to capture the deletion (by reading from the changefeed)
+		// and stop monitoring the cluster.
+		// TODO: Provide better communication between RP and Monitor
+		time.Sleep(time.Until(t.Add(time.Second * 20)))
```
Comment (darthhexx) [inline pkg/backend/openshiftcluster.go:193]: Adding 20 seconds to _every_ cluster delete is not a great customer experience to introduce for this edge case. Ideally we would follow the [recommended delete approach of using a TTLs](https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/change-feed-design-patterns#deletes), 10 seconds in our case, but due to use-cases where users are creating and tearing down clusters using the same name/RG/VNET, etc, we may run into issues when the TTL has not expired yet.

To limit the effect of this change, wouldn't it make sense to add a `deleteStartedAt := time.Now()` variable before `err = m.Delete(ctx)`, and then only sleep here if the delete has taken less than 10 seconds? Or 15 to be absolutely certain.

**Review by darthhexx — CHANGES_REQUESTED** [review]

File: pkg/backend/openshiftcluster.go (hunk 1 above)
Comment (darthhexx) [reply pkg/backend/openshiftcluster.go:193]: Ignore my first comment. I see that you have added a conditional sleep on the deletion start time.

**Review by darthhexx — APPROVED** [review]

File: pkg/backend/openshiftcluster.go (hunk 1 above)
Comment (darthhexx) [reply pkg/backend/openshiftcluster.go:193]: Ignore my previous comment, I see you have added a conditional sleep using the deletion start time.

**Review by petrkotas — APPROVED** [review]
@yehornaumenko perfect, this is exactly the good enough for now solution, we agreed upon. It will allow us to resolve some sev2.

In the future, we can improve the deletion mechanism.

File: docs/monitoring.md [hunk 2]
```diff
@@ -31,6 +31,12 @@ monitoring from inside the cluster as well as a complementary near-term goal.
   locally (like k8s list/watch).  At startup, the cosmos DB change feed returns
   the current state of all of the OpenShiftClusterDocuments; subsequently as
   OpenShiftClusterDocuments it returns the updated documents.
+* At the moment of writing, the change feed does not log record deletions. It logs
+  only changes. Deallocated clusters are deleted from the monitoring list only if
+  they were seen in the `DeletingProvisioningState` by the monitor.
+  The monitor reads the change feed every 10 seconds, so we should except exclude
```
Comment (facchettos) [inline docs/monitoring.md:37]:
```suggestion
  The monitor reads the change feed every 10 seconds, so we should expect exclude
```

**Review by facchettos — APPROVED** [review]
lgtm, little typos but that can be fixed later

**Pull request merged**